# Two-sided $z$-test

**date**
: 2021-04-17

**data**
: `practical-test.csv`

**module**
: `statsmodels.stats.weightstats.ztest`

**ref**
: Computer book B, Activity 23

**desc**
: Performing a two-sided $z$-test of a population mean.

## Import the modules

In [1]:
import src.load
from statsmodels.stats.weightstats import ztest

## Import the data

In [2]:
sample = src.load.practical_test()

In [3]:
sample.head()

Centre  Male  Female  Total
0         Aberdeen North  49.9    41.7   45.6
1  Aberdeen South (Cove)  50.9    43.9   47.2
2            Abergavenny  60.5    53.5   56.8
3            Aberystwyth  62.5    53.2   57.7
4                Airdrie  47.2    38.9   42.7

In [4]:
# declare two local vars for easier coding
total = sample["Total"]

## Scenario

The sample contains the pass rates for 316 UK driving practical test centres over the period April 2014 to March 2015.
The first column (`Centre`) lists the 316 test centres; the second and third columns (`Male` and `Female`) contain the pass rates (%) for males and females, respectively, at each centre; and the fourth column (`Total`) contains the overall pass rates (%) for each centre.

During the period April 2013 to March 2014, the national driving practical test pass rate was 47.1%.

## Hyotheses

Let the hypotheses be

$$
\begin{aligned}
  &H_{0} : \mu = 47.1\% \\
  &H_{1} : \mu \neq 47.1\%,
\end{aligned}
$$

where $\mu$ is the mean total pass rate for the driving practical test nationally across all UK test centres during the period April 2014 to March 2015.

## Method

We will use `statsmodels.stats.weightstats.ztest` to help with this analysis.

## Analysis

In [7]:
# get sample mean
total.mean()

49.630379746835445

In [5]:
# returns (z, p)
ztest(
    x1=total,
    value=47.1
)

(6.277491911091587, 3.440781584872767e-10)

## Discussion

Since $p <$ 0.01, there is strong evidence against the null hypothesis that the mean pass rate for the driving practical test nationally from April 2014 to March 2015 is the same as the national pass rate over the same period in the previous year.

Further, since $z \simeq$ 6.28, is positive, or equivalently because the sample mean, 49.629, is greater than the hypothesised population mean 47.1, the test suggests that the mean pass rate nationally over the period April 2014 to March 2015 is higher than the national pass rate over the same period in the previous year.

In [6]:
help(ztest)

Help on function ztest in module statsmodels.stats.weightstats:

ztest(x1, x2=None, value=0, alternative='two-sided', usevar='pooled', ddof=1.0)
    test for mean based on normal distribution, one or two samples
    
    In the case of two samples, the samples are assumed to be independent.
    
    Parameters
    ----------
    x1 : array_like, 1-D or 2-D
        first of the two independent samples
    x2 : array_like, 1-D or 2-D
        second of the two independent samples
    value : float
        In the one sample case, value is the mean of x1 under the Null
        hypothesis.
        In the two sample case, value is the difference between mean of x1 and
        mean of x2 under the Null hypothesis. The test statistic is
        `x1_mean - x2_mean - value`.
    alternative : str
        The alternative hypothesis, H1, has to be one of the following
    
           'two-sided': H1: difference in means not equal to value (default)
           'larger' :   H1: difference in means la